### PR TITLE
feat(backups): allow DR to remove previous backup first

### PR DESCRIPTION
**Pull request by badrAZ**
See vatesfr/xo-web#2157

### CODE REVIEW

File: src/xapi/index.js [hunk 1]
```diff
@@ -555,7 +555,11 @@ export default class Xapi extends XapiBase {
   } = {}) {
     // Fall back on local copy if possible.
     if (targetXapi === this) {
-      return this.copyVm(vmId, targetSrId, { nameLabel })
+      const vm = await this.copyVm(vmId, targetSrId, { nameLabel })
```
Comment (julien-f) [inline src/xapi/index.js]: Inline this variable.

File: src/xo-mixins/backups.js [hunk 2]
```diff
@@ -998,7 +998,27 @@ export default class {
     await Promise.all(promises)
   }
 
-  async rollingDrCopyVm ({vm, sr, tag, retention}) {
+  _drCopyVm = async (vm, sr, tag, sourceXapi, targetXapi) => {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Only use assignment for method declarations when you need them bound to the instance.

File: src/xo-mixins/backups.js [hunk 3]
```diff
@@ -998,7 +998,27 @@ export default class {
...
+    const copyName = `${vm.name_label}_${tag}_${safeDateFormat(new Date())}`
+    const data = await sourceXapi.remoteCopyVm(vm.$id, targetXapi, sr.$id, {
+      nameLabel: copyName
+    })
+
+    await targetXapi.addTag(data.vm.$id || data, 'Disaster Recovery')
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Why `data.vm.$id || data`?

File: src/xo-mixins/backups.js [hunk 4]
```diff
@@ -1017,21 +1037,18 @@ export default class {
...
-    const n = 1 - retention
-    await Promise.all(mapToArray(n ? olderCopies.slice(0, n) : olderCopies, vm =>
-      // Do not consider a failure to delete an old copy as a fatal error.
-      targetXapi.deleteVm(vm.$id)::pCatch(noop)
-    ))
+    if (deleteOldBackupsFirst) {
+      await this._removeOlderCopies(retention, olderCopies, targetXapi)
+      size = await this._drCopyVm(vm, sr, tag, sourceXapi, targetXapi)
+    } else {
+      size = await this._drCopyVm(vm, sr, tag, sourceXapi, targetXapi)
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Inline `_drCopyVm()` even if it means 2 tests instead of one.

File: src/xo-mixins/backups.js (hunk 3 above)
Comment (badrAZ) [inline src/xo-mixins/backups.js]: It's an error i will fix it.

File: src/xo-mixins/backups.js [hunk 5]
```diff
@@ -1017,21 +1017,40 @@ export default class {
...
+      size = data.size
+
+      await targetXapi.addTag(data.vm.$id, 'Disaster Recovery')
+
+      const n = 1 - retention
+      await Promise.all(mapToArray(n ? olderCopies.slice(0, n) : olderCopies, vm =>
+        // Do not consider a failure to delete an old copy as a fatal error.
+        targetXapi.deleteVm(vm.$id)::pCatch(noop)
+      ))
+    }
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Avoid code duplication: isolate deletion of the old backups in a function:

```js
if (deleteOldBackupsFirst) {
   await this._removeOld()
}

// DR here

if (!deleteFirst) {
  await this._removeOld()
}
```

File: src/xo-mixins/backups.js [hunk 6]
```diff
@@ -998,7 +998,16 @@ export default class {
     await Promise.all(promises)
   }
 
-  async rollingDrCopyVm ({vm, sr, tag, retention}) {
+  _removeOld = (retention, olderCopies, targetXapi) => {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Only use assignment when the method needs to be bound.

Comment (julien-f) [inline src/xo-mixins/backups.js]: Rename method to `_removeVms(xapi, vms)`.

File: src/xo-mixins/backups.js [hunk 7]
```diff
@@ -1017,18 +1026,20 @@ export default class {
     })
     const olderCopies = sortBy(vms, 'name_label')
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: `vmsToRemove = sortBy(vms, 'name_label').slice(0, 1 - retention)`

**Pull request merged**